postgres-native-sqldelight-driver/src/commonMain/kotlin/app/softwork/sqldelight/postgresdriver/ListenerSupport.kt

```diff
@@ -0,0 +1,71 @@
+package app.softwork.sqldelight.postgresdriver
+
+import kotlinx.cinterop.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+import libpq.*
+import platform.posix.*
+import kotlin.coroutines.*
+import kotlin.time.*
+
+public sealed interface ListenerSupport {
+    public sealed interface ScopedListenerSupport : ListenerSupport {
+        public val notificationScope: CoroutineScope
+    }
+
+    public companion object {
+        public fun Local(notificationScope: CoroutineScope): Local {
+            val notifications = MutableSharedFlow<String>()
+            return Local(notificationScope, notifications) { notifications.emit(it) }
+        }
+    }
+
+    public class Local(
+        override val notificationScope: CoroutineScope,
+        public val notifications: Flow<String>,
+        public val notify: suspend (String) -> Unit
+    ) : ScopedListenerSupport
+
+    public class Remote(
+        override val notificationScope: CoroutineScope,
+        public val notificationName: (String) -> String = { it }
+    ) : ScopedListenerSupport {
+        internal fun remoteListener(conn: CPointer<PGconn>): Flow<String> = channelFlow {
+            @OptIn(ExperimentalCoroutinesApi::class)
+            val selectorContext = newSingleThreadContext("WorkerSelectorManager")
+            val job = Job()
+            val coroutineContext = selectorContext + job
+
+            try {
+                val socket = PQsocket(conn)
+                check(socket >= 0) {
+                    "Error while connecting to the PostgreSql socket"
+                }
+                val pollfd: pollfd = memScoped { alloc() }
+                pollfd.fd = socket
+                val timeoutMs = 500
```

Comment:
Should this be configurable?

Reply:
Better, it should wait infinitely for new input and relaunching poll every 500ms. But using `timeoutMs = -1` needs cancelation support. And the only way to cancel poll is sending a packet to another file descriptor. At the end, this is implemented by Ktor. Unfortunately there is a bug: https://youtrack.jetbrains.com/issue/KTOR-5035/Remove-check-for-internal-class-in-Select, so I copied the implementation from Ktor.